pkg/service/auth_context.go

```diff
@@ -3,54 +3,119 @@ package service
 import (
 	"fmt"
 	"strings"
+	"sync"
 
 	"github.com/3scale/authorino/pkg/config"
+	"github.com/3scale/authorino/pkg/config/internal"
 
-	"golang.org/x/net/context"
 	auth "github.com/envoyproxy/go-control-plane/envoy/service/auth/v2"
+	"golang.org/x/net/context"
 )
 
 // AuthContext holds the context of each auth request, including the request itself (sent by the client),
 // the auth config of the requested API and the lists of identity verifications, metadata add-ons and
 // authorization policies, and their corresponding results after evaluated
 type AuthContext struct {
 	ParentContext *context.Context
-	Request *auth.CheckRequest
-	API *APIConfig
+	Request       *auth.CheckRequest
+	API           *APIConfig
+
+	Identity      map[*config.IdentityConfig]interface{}
+	Metadata      map[*config.MetadataConfig]interface{}
+	Authorization map[*config.AuthorizationConfig]interface{}
+}
 
-	Identity map[*config.IdentityConfig] interface{}
-	Metadata map[*config.MetadataConfig] interface{}
-	Authorization map[*config.AuthorizationConfig] interface{}
+// AuthObjectConfig provides an interface for APIConfig objects that implements a Call method
+type AuthObjectConfig interface {
+	Call(ctx internal.AuthContext) (interface{}, error)
+}
+
+func (authContext *AuthContext) getAuthObject(authObjConfig AuthObjectConfig) (interface{}, error) {
+	if ret, err := authObjConfig.Call(authContext); err != nil {
+		return nil, err
+	} else {
+		return ret, nil
+	}
+}
+
+// GetIDObject gets an Identity auth object given an Identity config.
+func (authContext *AuthContext) GetIDObject() error {
+	var wg sync.WaitGroup
+	var authObjError error
+	for _, config := range authContext.API.IdentityConfigs {
+		wg.Add(1)
+		var authObjCfg AuthObjectConfig = &config
+		go func() {
+			defer wg.Done()
+			if authObj, err := authContext.getAuthObject(authObjCfg); err != nil {
+				authObjError = err
```

Comment:
Using a shared variable to store the error is wrong. There may be concurrency issues and one "thread" could overwrite existing error to a `nil`.

Error handling in the goroutine context is done either by:
* Using an error channel where goroutines push errors and the main thread, after waiting for all to finish, checks for errors
* Use [errgroup](https://godoc.org/golang.org/x/sync/errgroup)